README.md
# Java Alchemy API

## About

Java Alchemy API is an interface to the [Alchemy API](http://www.alchemyapi.com/api/)

To access Alchemy's API you will need to [Register for an API Key](http://www.alchemyapi.com/api/register.html).

The Java SDK that Alchemy provides seemed very basic and incomplete to me so I created my own.

- They provide classes called "test" but they are examples.  They are not actual unit tests.

- The return object is always a dom Document.

- The source code format does not follow the Java standard.  For instance, method names begin with or contain
nothing but capital letters (e.g, GetInstanceFromFile, GET, POST).  Others are proper (e.g., getNodeValue)

- It gives the client the ability to specify the return format from the call to the service (either RDF or XML).
If you return objects there is no need for the client to specify or even care what the output format is.

### Be sure to see the [Changelog](https://github.com/likethecolor/Alchemy-API/blob/master/CHANGELOG.md)

### Note

A recent change (maven 1.1.0) has removed from the `Client` the loading of the API key from an external resource.

* `new Client()` does nothing

* `new Client(String)` expects the argument to be the API key

Removed:

* `new Client(String)` where the String is the classpath to a resource containing the API key

## Build

This project uses [Apache Maven](https://maven.apache.org/) to build.  To build and install into your local Maven repository:

        mvn clean install

## Maven Dependency

        <dependency>
            <groupId>com.likethecolor</groupId>
            <artifactId>alchemy</artifactId>
            <version>1.3.1</version>
        </dependency>

## Dependencies

You can look at the pom.xml to find the dependencies.  In short:

- [Apache Commons Lang](https://commons.apache.org/lang/)

- [Simple Logging Facade for Java (SLF4J)](http://www.slf4j.org/)

- [JSON](http://www.json.org/java/index.html)

Note that for testing the slf4j-log4j12 artifact is a dependency (scope: test).  Logging is available when running unit tests if needed.

## Code Examples

### Using Default Parameters

        final String apiKey = "0a1b2c3d4e5f6g7h8i9j0a1b2c3d4e5f6g7h8i9j";
        final Client client = new Client(apiKey);

        OR

        final Client client = new Client();
        client.setAPIKey(apiKey);

        final AbstractCall<AuthorAlchemyEntity> authorCall = new AuthorCall(new CallTypeUrl("http://www.politico.com/blogs/media/2012/02/detroit-news-ed-upset-over-romney-edit-115247.html"));
        final Response response = client.call(authorCall);
        System.out.println(response);
        System.out.println();

which is slightly more verbose version of this:

        final AbstractCall<AuthorAlchemyEntity> authorCall = new AuthorCall(new CallTypeUrl("http://www.politico.com/blogs/media/2012/02/detroit-news-ed-upset-over-romney-edit-115247.html"));
        System.out.println(client.call(authorCall));
        System.out.println();

More examples:

        final AbstractCall<LanguageAlchemyEntity> languageCall = new LanguageCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
        System.out.println(client.call(languageCall).toString(ToStringStyle.MULTI_LINE_STYLE));

        final AbstractCall<MicroformatAlchemyEntity> microformatCall = new MicroformatDataCall(new CallTypeUrl("http://microformats.org/wiki/hcard"));
        System.out.println(client.call(microformatCall).toString(ToStringStyle.MULTI_LINE_STYLE));

        final AbstractCall<TaxonomyAlchemyEntity> rankedTaxonomyCall = new RankedTaxonomyCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
        System.out.println(client.call(rankedTaxonomyCall).toString(ToStringStyle.MULTI_LINE_STYLE));

        final AbstractCall<ConceptAlchemyEntity> rankedConceptsCall = new RankedConceptsCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
        System.out.println(client.call(rankedConceptsCall).toString(ToStringStyle.MULTI_LINE_STYLE));

        final AbstractCall<KeywordAlchemyEntity> rankedKeywordsCall = new RankedKeywordsCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
        System.out.println(client.call(rankedKeywordsCall).toString(ToStringStyle.MULTI_LINE_STYLE));

        final AbstractCall<HeaderAlchemyEntity> rawTextCall = new RawTextCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
        System.out.println(client.call(rawTextCall).toString(ToStringStyle.MULTI_LINE_STYLE));

        final AbstractCall<RelationAlchemyEntity> relationsCall = new RelationsCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
        System.out.println(client.call(relationsCall).toString(ToStringStyle.MULTI_LINE_STYLE));

        final AbstractCall<SentimentAlchemyEntity> sentimentCall = new SentimentCall(new CallTypeUrl("http://www.answers.com/article/3b7a1506797deda268075f0886eac4d8"));
        System.out.println(client.call(rawTextCall).toString(ToStringStyle.MULTI_LINE_STYLE);

        final AbstractCall<HeaderAlchemyEntity> textSentimentCall = new TextCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
        System.out.println(client.call(textSentimentCall).toString(ToStringStyle.MULTI_LINE_STYLE));

        final AbstractCall<TitleAlchemyEntity> titleCall = new TitleCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
        System.out.println(client.call(titleCall).toString(ToStringStyle.MULTI_LINE_STYLE));
        
        # v 1.2.0
        final AbstractCall<ImageEntity> imageCall = new ImageCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
        System.out.println(client.call(imageCall).toString(ToStringStyle.MULTI_LINE_STYLE));


### Setting Optional Values

        final TargetedSentimentParams targetedSentimentParams = new TargetedSentimentParams();
        targetedSentimentParams.setTarget("Schrenker");

        final AbstractCall<SentimentAlchemyEntity> targetedSentimentCall = new TargetedSentimentCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"), targetedSentimentParams);
        System.out.println(client.call(targetedSentimentCall));

        final NamedEntityParams namedEntityParams = new NamedEntityParams();
        namedEntityParams.setIsCoreference(true);
        namedEntityParams.setIsDisambiguate(true);
        namedEntityParams.setIsLinkedData(true);
        namedEntityParams.setIsQuotations(true);
        namedEntityParams.setIsSentiment(true);
        namedEntityParams.setIsShowSourceText(true);
        namedEntityParams.setLanguage(Language.ENGLISH);

        final AbstractCall<NamedEntityAlchemyEntity> rankedNamedEntitiesCall = new RankedNamedEntitiesCall(new CallTypeUrl("http://www.politico.com/blogs/media/2012/02/detroit-news-ed-upset-over-romney-edit-115247.html"), namedEntityParams);
        final Response<NamedEntityAlchemyEntity> rankedNamedEntitiesResponse = client.call(rankedNamedEntitiesCall);
        System.out.println(rankedNamedEntitiesResponse.toString(ToStringStyle.MULTI_LINE_STYLE));

### Complete Example

        final String apiKey = "0a1b2c3d4e5f6g7h8i9j0a1b2c3d4e5f6g7h8i9j";
        final Client client = new Client(apiKey);

        // Big Blue's Bet on Big Data
        final AbstractCall<SentimentAlchemyEntity> sentimentCall = new SentimentCall(new CallTypeUrl("http://www.answers.com/article/3b7a1506797deda268075f0886eac4d8"));
        final Response<SentimentAlchemyEntity> sentimentResponse = client.call(sentimentCall);

        System.out.println("Language: " + sentimentResponse.getLanguage());
        System.out.println("Status: " + sentimentResponse.getStatus());
        System.out.println("Status Info: " + sentimentResponse.getStatusInfo());
        System.out.println("Text: " + sentimentResponse.getText());
        System.out.println("Usage: " + sentimentResponse.getUsage());
        System.out.println("URL: " + sentimentResponse.getURL());

        SentimentAlchemyEntity entity;
        final Iterator<SentimentAlchemyEntity> iter = sentimentResponse.iterator();
        while(iter.hasNext()) {
          entity = iter.next();
          System.out.println("isMixed: " + (entity.isMixed() ? "true" : "false"));
          System.out.println("Score: " + entity.getScore());
          System.out.println("Type: " + entity.getType());
        }

## Thanks

Thank you to the following for providing a second set of eyes and helping me
make this project better.

- [Ainur Minibaev](rapper01@list.ru) [ainurminibaev] (https://github.com/ainurminibaev)
- [Amir Raminfar](findamir@gmail.com) [amir20](https://github.com/amir20)
- [bellorap](https://github.com/bellorap)
- [Yair Liberzon](yair.liberzon@zoomd.com) and [Sivan Ravid](sivan.ravid@zoomd.com)
- [Gursev Singh Kaira] (https://github.com/gursev)
- [LucaMarzialetti] (https://github.com/LucaMarzialetti)


## Authors

Dan Brown (<dan@likethecolor.com>)

## External API

- [Alchemy API](http://www.alchemyapi.com/api/)

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
# Changelog

## 1.3.1 - Sep 09, 2016

- bug-22: can now specify language in the params [LucaMarzialetti](https://github.com/LucaMarzialetti)

## 1.3.0 - Jun 20, 2016

- [support] (support@alchemyapi.com) provided a new url for the invalid-permissions-for-call error

## 1.2.1 - Jun 11, 2016

- Thanks to [Gursev Singh Kaira] (https://github.com/gursev) asking that the call to AlchemyAPI service use https

## 1.2.0 - Dec 28, 2015

- added Ranked Image Keywords call provided by [ainurminibaev](https://github.com/ainurminibaev)

## 1.1.6 - Mar 12, 2015

-  bug-13: reduced the min length on text that can be sent from 5 to 2

## 1.1.5 - Nov 15, 2014

-  bug-11: [Yair Liberzon] (yair.liberzon@zoomd.com) and [Sivan Ravid](sivan.ravid@zoomd.com) noted that the log messages still show up when the Alchemy service returns the non-standard JSON boolean.

 The only differences between 1.1.4 and this version is that:

    - the log levels for those messages changed from WARN to DEBUG
    - java.source.version changed from 1.6 to 1.7 (might as well be current)
    - java.target.version changed from 1.6 to 1.7 (might as well be current)

## 1.1.4 - Oct 15, 2014

-  bug-11: Thanks to [Yair Liberzon] (yair.liberzon@zoomd.com) and [Sivan Ravid](sivan.ravid@zoomd.com) for reporting this bug
  - AlchemyAPI code expects standard JSON boolean value true/false, the AlchemyAPI service returns yes/no

## 1.1.3 - Oct 09, 2014

-  bug-10: Thanks to [bellorap](https://github.com/bellorap) for discovering this bug.
  - Client#post(String,String,Params) now throws IOException instead of swallowing it.

## 1.1.2 - May 30, 2014

-  Ranked Categories - not supported by AlchemyApi? ([#8](https://github.com/likethecolor/Alchemy-API/issues/8))

## 1.1.1 - Oct 02, 2013

- Removed unnecessary IOException from Client (thanks to [Amir Raminfar](https://github.com/amir20))
- Unchecked assignment when doing client.call  ([#6](https://github.com/likethecolor/Alchemy-API/issues/6)) ([#7](https://github.com/likethecolor/Alchemy-API/issues/7))  (thanks to [Amir Raminfar](https://github.com/amir20))

## 1.1.0 - Oct 01, 2013

- setAPIKey should be private ([#5](https://github.com/likethecolor/Alchemy-API/issues/5))  (thanks to [Amir Raminfar](https://github.com/amir20))

## 1.0.1 - Jan 07, 2013
- OutputStatusValidator logs the json string returned from AlchemyAPI service ([#2](https://github.com/likethecolor/Alchemy-API/issues/2))
- AbstractParser sets status instead of status info ([#3](https://github.com/likethecolor/Alchemy-API/issues/3))


src/main/java/com/likethecolor/alchemy/Main.java
/**
 * Original Author: Dan Brown <dan@likethecolor.com>
 * Copyright 2015 Dan Brown <dan@likethecolor.com>
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.likethecolor.alchemy;

import com.likethecolor.alchemy.api.Client;
import com.likethecolor.alchemy.api.call.AbstractCall;
import com.likethecolor.alchemy.api.call.AuthorCall;
import com.likethecolor.alchemy.api.call.ImageCall;
import com.likethecolor.alchemy.api.call.LanguageCall;
import com.likethecolor.alchemy.api.call.MicroformatDataCall;
import com.likethecolor.alchemy.api.call.RankedConceptsCall;
import com.likethecolor.alchemy.api.call.RankedKeywordsCall;
import com.likethecolor.alchemy.api.call.RankedNamedEntitiesCall;
import com.likethecolor.alchemy.api.call.RankedTaxonomyCall;
import com.likethecolor.alchemy.api.call.RawTextCall;
import com.likethecolor.alchemy.api.call.RelationsCall;
import com.likethecolor.alchemy.api.call.SentimentCall;
import com.likethecolor.alchemy.api.call.TargetedSentimentCall;
import com.likethecolor.alchemy.api.call.TextCall;
import com.likethecolor.alchemy.api.call.TitleCall;
import com.likethecolor.alchemy.api.call.type.CallTypeUrl;
import com.likethecolor.alchemy.api.entity.AuthorAlchemyEntity;
import com.likethecolor.alchemy.api.entity.ConceptAlchemyEntity;
import com.likethecolor.alchemy.api.entity.HeaderAlchemyEntity;
import com.likethecolor.alchemy.api.entity.ImageEntity;
import com.likethecolor.alchemy.api.entity.KeywordAlchemyEntity;
import com.likethecolor.alchemy.api.entity.LanguageAlchemyEntity;
import com.likethecolor.alchemy.api.entity.MicroformatAlchemyEntity;
import com.likethecolor.alchemy.api.entity.NamedEntityAlchemyEntity;
import com.likethecolor.alchemy.api.entity.RelationAlchemyEntity;
import com.likethecolor.alchemy.api.entity.Response;
import com.likethecolor.alchemy.api.entity.SentimentAlchemyEntity;
import com.likethecolor.alchemy.api.entity.TaxonomyAlchemyEntity;
import com.likethecolor.alchemy.api.entity.TitleAlchemyEntity;
import com.likethecolor.alchemy.api.params.Language;
import com.likethecolor.alchemy.api.params.NamedEntityParams;
import com.likethecolor.alchemy.api.params.TargetedSentimentParams;
import org.apache.commons.lang.builder.ToStringStyle;

import java.io.IOException;
import java.util.Iterator;

public class Main {
  public static void main(String[] args) throws IOException {
    final Client client = new Client("random0alchemy0api0key");

    AbstractCall<AuthorAlchemyEntity> authorCall = new AuthorCall(new CallTypeUrl("http://www.politico.com/blogs/media/2012/02/detroit-news-ed-upset-over-romney-edit-115247.html"));
    System.out.println(client.call(authorCall));
    System.out.println();

    authorCall = new AuthorCall(new CallTypeUrl("http://www.politico.com/blogs/media/2012/02/detroit-news-ed-upset-over-romney-edit-115247.html"));
    Response response = client.call(authorCall);
    System.out.println(response);
    System.out.println();
//
////More examples:
//
    AbstractCall<LanguageAlchemyEntity> languageCall = new LanguageCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
    System.out.println(client.call(languageCall).toString(ToStringStyle.MULTI_LINE_STYLE));
    System.out.println();

    AbstractCall<MicroformatAlchemyEntity> microformatCall = new MicroformatDataCall(new CallTypeUrl("http://microformats.org/wiki/hcard"));
    System.out.println(client.call(microformatCall).toString(ToStringStyle.MULTI_LINE_STYLE));
    System.out.println();

		AbstractCall<TaxonomyAlchemyEntity> rankedTaxonomyCall = new RankedTaxonomyCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
		System.out.println(client.call(rankedTaxonomyCall).toString(ToStringStyle.MULTI_LINE_STYLE));
		System.out.println();

		AbstractCall<ConceptAlchemyEntity> rankedConceptsCall = new RankedConceptsCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
    System.out.println(client.call(rankedConceptsCall).toString(ToStringStyle.MULTI_LINE_STYLE));
    System.out.println();

    AbstractCall<KeywordAlchemyEntity> rankedKeywordsCall = new RankedKeywordsCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
    System.out.println(client.call(rankedKeywordsCall).toString(ToStringStyle.MULTI_LINE_STYLE));
    System.out.println();

    AbstractCall<HeaderAlchemyEntity> rawTextCall = new RawTextCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
    System.out.println(client.call(rawTextCall).toString(ToStringStyle.MULTI_LINE_STYLE));
    System.out.println();

    AbstractCall<RelationAlchemyEntity> relationsCall = new RelationsCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
    System.out.println(client.call(relationsCall).toString(ToStringStyle.MULTI_LINE_STYLE));
    System.out.println();

    // Big Blue's Bet on Big Data
    AbstractCall<SentimentAlchemyEntity> sentimentCall = new SentimentCall(new CallTypeUrl("http://www.answers.com/article/3b7a1506797deda268075f0886eac4d8"));
    Response<SentimentAlchemyEntity> sentimentResponse = client.call(sentimentCall);

    System.out.println("Language: " + sentimentResponse.getLanguage());
    System.out.println("Status: " + sentimentResponse.getStatus());
    System.out.println("Status Info: " + sentimentResponse.getStatusInfo());
    System.out.println("Text: " + sentimentResponse.getText());
    System.out.println("Usage: " + sentimentResponse.getUsage());
    System.out.println("URL: " + sentimentResponse.getURL());

    SentimentAlchemyEntity entity;
    Iterator<SentimentAlchemyEntity> iter = sentimentResponse.iterator();
    while(iter.hasNext()) {
      entity = iter.next();
      System.out.println("isMixed: " + (entity.isMixed() ? "true" : "false"));
      System.out.println("Score: " + entity.getScore());
      System.out.println("Type: " + entity.getType());
    }

    sentimentCall = new SentimentCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
    System.out.println(client.call(sentimentCall));
    System.out.println();

    AbstractCall<HeaderAlchemyEntity> textSentimentCall = new TextCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
    System.out.println(client.call(textSentimentCall).toString(ToStringStyle.MULTI_LINE_STYLE));
    System.out.println();

    AbstractCall<TitleAlchemyEntity> titleCall = new TitleCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
    System.out.println(client.call(titleCall).toString(ToStringStyle.MULTI_LINE_STYLE));
    System.out.println();
//
////Setting Optional Values
////-----------------------
    final TargetedSentimentParams params = new TargetedSentimentParams();
    params.setTarget("Schrenker");

    AbstractCall<SentimentAlchemyEntity> targetedSentimentCall = new TargetedSentimentCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"), params);
    System.out.println(client.call(targetedSentimentCall));


    final NamedEntityParams namedEntityParams = new NamedEntityParams();
    namedEntityParams.setIsCoreference(true);
    namedEntityParams.setIsDisambiguate(true);
    namedEntityParams.setIsLinkedData(true);
    namedEntityParams.setIsQuotations(true);
    namedEntityParams.setIsSentiment(true);
    namedEntityParams.setIsShowSourceText(true);
    namedEntityParams.setLanguage(Language.ENGLISH);

    AbstractCall<NamedEntityAlchemyEntity> rankedNamedEntitiesCall = new RankedNamedEntitiesCall(new CallTypeUrl("http://www.politico.com/blogs/media/2012/02/detroit-news-ed-upset-over-romney-edit-115247.html"), namedEntityParams);
    Response<NamedEntityAlchemyEntity> rankedNamedEntitiesResponse = client.call(rankedNamedEntitiesCall);
    System.out.println(rankedNamedEntitiesResponse.toString(ToStringStyle.MULTI_LINE_STYLE));

    AbstractCall<ImageEntity> imageCall = new ImageCall(new CallTypeUrl("http://www.cnn.com/2009/CRIME/01/13/missing.pilot/index.html"));
    System.out.println(client.call(imageCall).toString(ToStringStyle.MULTI_LINE_STYLE));
    System.out.println();
  }
}
